AGP 8.0 build.gradle namespace and aidl buildFeature requirement added

onuralpszr opened the pull request:
Hello,

Android Gradle Plugin version 8.0 is asking for namespace. This is become mandatory and after I update my AGP to 8.0, I got this error 


```
Namespace not specified. Please specify a namespace in the module's build.gradle file like so:

android {
    namespace 'com.example.namespace'
}

If the package attribute is specified in the source AndroidManifest.xml, it can be migrated automatically to the namespace value in the build.gradle file using the AGP Upgrade Assistant; please refer to https://developer.android.com/studio/build/agp-upgrade-assistant for more information.
```

This change fix this future releases. However I am not sure how opencv wants to user namespace I used "org.opencv" if there is a different namespace please let me know so I can changed that too. Also should I add namepsace into "opencv/modules/java/android_sdk/android_gradle_lib/build.gradle" here ?


### Sources

Android developer link: https://developer.android.com/studio/preview/features#namespace-dsl
Issue Tracker Google: https://issuetracker.google.com/issues/191813691?pli=1#comment19






### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

asmorkalov commented:
OpenCV still uses very old versions of NDK, SDK and Gradle for pre-commit testing and releases. Gradle update at least to 7.3.3 and Android plugin update at least to 7.2 is required. Thanks a lot for the contribution! We will return back to your patch after build infra update.

onuralpszr commented:
Using gradle 7.x basically fix the problem. İn gradle 8.0 it become mandatory, it does fix that,as well.

opencv-alalek commented on cmake/android/android_gradle_projects.cmake:
```diff
@@ -82,6 +82,9 @@ foreach(fname ${GRADLE_WRAPPER_FILES})
 endforeach()
 
+# set build.gradle namespace 
+set(OPENCV_NAMESPACE "org.opencv" CACHE STRING "OpenCV Namespace string")
```
Need to use version check to avoid breaking of existed build configurations:

```
if(... GRADLE VERSION CHECK ...)
    ocv_update(OPENCV_NAMESPACE_DECLARATION "namespace 'org.opencv'")
else()
    ocv_update(OPENCV_NAMESPACE_DECLARATION "")
endif()
```

and below:

```.diff
 android {
+    @OPENCV_NAMESPACE_DECLARATION@
```

onuralpszr replied:
I made those changes accordingly so please check it out.

onuralpszr commented:
I don't wanna open up a new pull request and I can If needs to be but we also need to add

```
    buildFeatures {
        aidl true
    }
```

https://developer.android.com/build/releases/gradle-plugin#default-changes

Inside of the gradle file because new default become false, gradle 7 and 5 was "true" so it won't effect them and It should be  fine for people who start brand new project so they can avoid aidl error

onuralpszr commented:
@asmorkalov  I was also looking for JDK version but could not found it. 

https://developer.android.com/build/releases/gradle-plugin#compatibility-8-0-0 also set JDK minimum to 17. As far as I know opencv uses jdk 11 but user needs to set them to 17 for newer gradle 8.0+ for old versions they can leave as it is. I already tested JDK 17 and it works fine. 

For gradle 8.0 
```
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_17
        targetCompatibility JavaVersion.VERSION_17
    }
    
```
    
   For older version I just need to VERSION_11 (based on opencv minimum) I would like to add this so It can build without error. So can we add at least something similar to gradle or add  hard coded VERSION_11 and advise users to change on of top gradle readme like similar to set SDK versions

asmorkalov commented:
CI reports issue:
```
cmake/android/android_gradle_projects.cmake:85: trailing whitespace.
+# set build.gradle namespace 
```

asmorkalov commented on cmake/android/android_gradle_projects.cmake:
```diff
@@ -82,6 +82,14 @@ foreach(fname ${GRADLE_WRAPPER_FILES})
+  ocv_update(OPENCV_NAMESPACE_DECLARATION "namespace 'org.opencv'")
+endif()
+
+
```
No need extra space here.

onuralpszr replied:
I removed it.

opencv-alalek commented on cmake/android/android_gradle_projects.cmake:
```diff
@@ -82,6 +82,13 @@ foreach(fname ${GRADLE_WRAPPER_FILES})
   install(FILES "${OpenCV_SOURCE_DIR}/platforms/android/gradle-wrapper/${fname}" DESTINATION "${ANDROID_INSTALL_SAMPLES_DIR}/${__dir}" COMPONENT samples ${__permissions})
 endforeach()
 
+# set build.gradle namespace 
+if(GRADLE_VERSION STREQUAL "5.6.4")
+  ocv_update(OPENCV_NAMESPACE_DECLARATION "")
+else()
+  ocv_update(OPENCV_NAMESPACE_DECLARATION "namespace 'org.opencv'")
+endif()
```
What would happen if `GRADLE_VERSION` is 5.6.3?

onuralpszr replied:
Namescape came with AGP 7.3 (https://developer.android.com/build/releases/gradle-plugin) so is it okay to set minimum 7.3.0 and any higher should be fine. Any lower version basically won't use it. That include 5.6.3 or any X version lower then 7.3.0

```
# set build.gradle namespace 
if(GRADLE_VERSION VERSION VERSION_GREATER_EQUAL "7.3.0")
  ocv_update(OPENCV_NAMESPACE_DECLARATION "namespace 'org.opencv'")
else()
  ocv_update(OPENCV_NAMESPACE_DECLARATION "")
endif()

```

opencv-alalek replied:
Right, condition should use GE version check instead of strong equality.

onuralpszr replied:
Did you mean this ? 

if(GRADLE_VERSION VERSION VERSION_GREATER "7.2.9").

opencv-alalek replied:
Both cases should do the similar thing:

```
if(GRADLE_VERSION VERSION_GREATER_EQUAL "7.3.0")
if(GRADLE_VERSION VERSION_GREATER "7.2.9")
```

The first option is clear however.

---

BTW, "gradle plugin version" is not the same as "gradle version".
Plugin version is specified through `ANDROID_GRADLE_PLUGIN_VERSION`

onuralpszr replied:
I changed with AGP and push it.

opencv-alalek commented on cmake/android/android_gradle_projects.cmake:
```diff
@@ -82,6 +82,13 @@ foreach(fname ${GRADLE_WRAPPER_FILES})
   install(FILES "${OpenCV_SOURCE_DIR}/platforms/android/gradle-wrapper/${fname}" DESTINATION "${ANDROID_INSTALL_SAMPLES_DIR}/${__dir}" COMPONENT samples ${__permissions})
 endforeach()
 
+# set build.gradle namespace 
```
Please fix whitespace issues:

https://pullrequest.opencv.org/buildbot/builders/precommit_docs/builds/102951/steps/whitespace%20opencv/logs/stdio

(remove extra spaces in EOL)

onuralpszr replied:
I didn't see it, it is done.

onuralpszr commented:
@opencv-alalek  can AGP version set to at least 4.2 ? for "aidl" it is gonna be fail because of that. And 3.2.1 is really old ?

onuralpszr commented:
> @asmorkalov I was also looking for JDK version but could not found it.
> 
> https://developer.android.com/build/releases/gradle-plugin#compatibility-8-0-0 also set JDK minimum to 17. As far as I know opencv uses jdk 11 but user needs to set them to 17 for newer gradle 8.0+ for old versions they can leave as it is. I already tested JDK 17 and it works fine.
> 
> For gradle 8.0
> 
> ```
>     compileOptions {
>         sourceCompatibility JavaVersion.VERSION_17
>         targetCompatibility JavaVersion.VERSION_17
>     }
>     
> ```
> 
> For older version I just need to VERSION_11 (based on opencv minimum) I would like to add this so It can build without error. So can we add at least something similar to gradle or add hard coded VERSION_11 and advise users to change on of top gradle readme like similar to set SDK versions

@opencv-alalek  I also need suggestion for this one too because of AGP(8.0) again.

opencv-alalek commented:
Perhaps we need to introduce some variable for Java version (similar to `ANDROID_GRADLE_PLUGIN_VERSION`).
E.g., `ANDROID_GRADLE_JAVA_VERSION` (prefer digits only to use version comparison).

opencv-alalek commented on cmake/android/android_gradle_projects.cmake:
```diff
@@ -82,6 +82,13 @@ foreach(fname ${GRADLE_WRAPPER_FILES})
 endforeach()
 
+# set build.gradle namespace
+if(ANDROID_GRADLE_PLUGIN_VERSION VERSION_GREATER_EQUAL "7.3.0")
```
> VERSION_GREATER_EQUAL

Some old CMake versions doesn't support that. Please use check in this form:

```
if(NOT (ANDROID_GRADLE_PLUGIN_VERSION VERSION_LESS "7.3.0"))
```

opencv-alalek commented on cmake/android/android_gradle_projects.cmake:
```diff
@@ -82,6 +82,13 @@ foreach(fname ${GRADLE_WRAPPER_FILES})
 
+# set build.gradle namespace
+if(ANDROID_GRADLE_PLUGIN_VERSION VERSION_GREATER_EQUAL "7.3.0")
+  ocv_update(OPENCV_NAMESPACE_DECLARATION "namespace 'org.opencv'")
```
`OPENCV_NAMESPACE_DECLARATION` -> `OPENCV_ANDROID_NAMESPACE_DECLARATION` is preferable to avoid possible conflicts.

onuralpszr commented:
@opencv-alalek  both changed added.

opencv-alalek commented:
@onuralpszr Thanks! Triggered CI with latest changes.

Please take a look on the comment above about `ANDROID_GRADLE_JAVA_VERSION` proposal for `JavaVersion` description (if required).

onuralpszr commented:
> Perhaps we need to introduce some variable for Java version (similar to `ANDROID_GRADLE_PLUGIN_VERSION`). E.g., `ANDROID_GRADLE_JAVA_VERSION` (prefer digits only to use version comparison).

Sure, based on suggestion here is how is it look like 

Cmake changes


```cmake
set(ANDROID_GRADLE_JAVA_VERSION 11 CACHE STRING "Android Gradle Java version")
message(STATUS "Android Gradle Java version: ${ANDROID_GRADLE_JAVA_VERSION}")
```

```cmake
# set android gradle java version in build.gradle
if(NOT (ANDROID_GRADLE_PLUGIN_VERSION VERSION_LESS "8.0.0"))
  # AGP-8.0 requires a minimum JDK version of JDK18
  ocv_update(ANDROID_GRADLE_JAVA_VERSION "18")
else()
  ocv_update(ANDROID_GRADLE_JAVA_VERSION "${ANDROID_GRADLE_JAVA_VERSION}")
endif()
```

Gradle part

```gradle
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_@ANDROID_GRADLE_JAVA_VERSION@
        targetCompatibility JavaVersion.VERSION_@ANDROID_GRADLE_JAVA_VERSION@
    }
```

opencv-alalek commented:
> how is it look like

Looks good.
Please apply changes and we will check them on existed build configurations.

onuralpszr commented:
> > how is it look like
> 
> Looks good. Please apply changes and we will check them on existed build configurations.

I added JDK sections please trigger CI and one more thing,

I removed "aidl" temporally part because it needs to be added via condition because current AGP in CI not support "buildFeatures" which failed on last also failed on last CI trigger.

```gradle
    buildFeatures {
        aidl true
    }
```

onuralpszr commented:
@opencv-alalek  

I set to JDK 1.8 based on CI and I also fix the typo "AGP8.0 JDK version 17 not 18 sorry about that. I fix both and please trigger the CI

opencv-alalek commented:
BTW, there are "config" files for Android packages here: https://github.com/opencv/opencv/tree/4.x/platforms/android

It makes sense to create new one configuration file which describes your tested configuration (where to pass required CMake parameters (like `ANDROID_GRADLE_JAVA_VERSION`) through `cmake_vars` dictionary).

onuralpszr commented on cmake/android/android_gradle_projects.cmake:
```diff
@@ -2,6 +2,9 @@
 set(ANDROID_GRADLE_PLUGIN_VERSION "3.2.1" CACHE STRING "Android Gradle Plugin version")
 message(STATUS "Android Gradle Plugin version: ${ANDROID_GRADLE_PLUGIN_VERSION}")
 
+set(ANDROID_GRADLE_JAVA_VERSION "1_8" CACHE STRING "Android Gradle Java version")
```
I pushed the changes

onuralpszr commented:
> BTW, there are "config" files for Android packages here: https://github.com/opencv/opencv/tree/4.x/platforms/android
> 
> It makes sense to create new one configuration file which describes your tested configuration (where to pass required CMake parameters (like `ANDROID_GRADLE_JAVA_VERSION`) through `cmake_vars` dictionary).

I am not sure I understand this, because problem is based on AGP/Gradle/Java but not "cmake" ? I didn't understand that adding another config file based on "which ndk" version ?

opencv-alalek reviewed (approved):
LGTM :+1:

onuralpszr commented:
> > > how is it look like
> > 
> > 
> > Looks good. Please apply changes and we will check them on existed build configurations.
> 
> I added JDK sections please trigger CI and one more thing,
> 
> I removed "aidl" part temporally because it needs to be added via condition because current AGP in CI not support "buildFeatures" which failed on last CI trigger.
> 
> ```groovy-gradle
>     buildFeatures {
>         aidl true
>     }
> ```

@opencv-alalek  we still need to add "aidl" part without that it will fail on version 8.0.0 and above.  

```cmake
# set android gradle java version in build.gradle and set aidl config
if(NOT (ANDROID_GRADLE_PLUGIN_VERSION VERSION_LESS "8.0.0"))
  # AGP-8.0 requires a minimum JDK version of JDK17
  ocv_update(ANDROID_GRADLE_JAVA_VERSION_INIT "17")
  # AGP-8.0 default aidl is false, OpenCV require aidl config to be true for compile
  ocv_update(ANDROID_GRADLE_BUILD_FEATURE_AIDL "buildFeatures { aidl true }")
else()
  ocv_update(ANDROID_GRADLE_JAVA_VERSION_INIT "1_8")
  ocv_update(ANDROID_GRADLE_BUILD_FEATURE_AIDL "")
endif()
```

onuralpszr commented:
> LGTM +1

Thank you for reviews.

0xMaty commented:
The solution haven't quite worked for me. It worked after I modified gradle script as follows:
```
buildFeatures {
    aidl true
    buildConfig true
}
````

onuralpszr commented:
> The solution haven't quite worked for me. It worked after I modified gradle script as follows:
> 
> ```
> buildFeatures {
>     aidl true
>     buildConfig true
> }
> ```

For building SDK or using SDK ?

Because for usagei I made a template project for a while ago and I am using standard android project and you can see in this config and patch and you even try yourself it is works. I also made a CI  so we can see it does build.

https://github.com/onuralpszr/CvCamera-Mobile/blob/main/patches/cv_build_gradle.patch
https://github.com/onuralpszr/CvCamera-Mobile/tree/main

Do you have a project shows you it is not builds ?

0xMaty commented:
When I imported OpenCV and tried to build my project. It failed on missing BuildConfig class. Once I added the `buildConfig true` - my project compiled and I was able to successfully initialize OpenCV.

nando8082 commented:
When I imported OpeCV in my android studio project (Android Studio Giraffe | 2022.3.1 Patch 2) and I tried to buil my project appear the next error
Plugin [id: 'com.android.application', version: '8.1.2', apply: false] was not found in any of the following sources:

BrokenSoul228 commented:
Hi, I added module OpenCV to my android project on kotlin, try to invoke VideoCapture, but after build take issue "could not find "libopencv_java490.so"".  I have found "libopencv_java4.so" in my project. Can anyone help me with my problem?
